### PR TITLE
llama : fix codeshell support (#8250)

### PR DESCRIPTION
**Pull request by hankeke303**
- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [x] Low
  - [ ] Medium
  - [ ] High

Hi, @ggerganov 

We submitted support for CodeShell in PR #5016, but with the update of `llama.cpp`, there were some issues with CodeShell support (We have reported this in Issue #8250). This PR aims to sync with the latest version of the repository and fix these issues.

### CODE REVIEW

**Review by ggerganov — APPROVED** [review]

**Comment by ggerganov** [comment]
After resolving the conflict from https://github.com/ggerganov/llama.cpp/pull/8609, we can merge

**Comment by hankeke303** [comment]
> After resolving the conflict from #8609, we can merge

@ggerganov We have resolve this conflict. Please review and apply it again.

Thanks for taking some time to review it!

File: src/llama.cpp [hunk 1]
```diff
@@ -5515,6 +5515,9 @@ static void llm_load_vocab(
             } else if (
                 tokenizer_pre == "jais") {
                 vocab.type_pre = LLAMA_VOCAB_PRE_TYPE_JAIS;
+            } else if (
+                tokenizer_pre == "codeshell") {
+                vocab.type_pre = LLAMA_VOCAB_PRE_TYPE_CODESHELL;
```
Comment (ggerganov) [inline src/llama.cpp:5520]: Move this block after `smollm` below to respect the enum order

File: src/llama.cpp [hunk 2]
```diff
@@ -15547,6 +15550,7 @@ struct llm_tokenizer_bpe {
             case LLAMA_VOCAB_PRE_TYPE_STARCODER:
             case LLAMA_VOCAB_PRE_TYPE_REFACT:
             case LLAMA_VOCAB_PRE_TYPE_COMMAND_R:
+            case LLAMA_VOCAB_PRE_TYPE_CODESHELL:
```
Comment (ggerganov) [inline src/llama.cpp:15553]: Move one line below the `smollm`

**Comment by hankeke303** [comment]
@ggerganov Sorry for not noticing the order before, we have corrected it.

**Pull request merged**